pydantic_evals/pydantic_evals/datasets.py

```diff
@@ -0,0 +1,376 @@
+            elif len(args := getattr(c, '__args__', ())) == 3:
+                return args
+        raise ValueError(f'Could not determine the generic parameters for {cls}')
+
+    @classmethod
+    def assessment(
+        cls, f: AssessmentFunction[InputsT, OutputT, MetadataT]
+    ) -> AssessmentFunction[InputsT, OutputT, MetadataT]:
+        """Decorator that registers an assessment function in the class-specific registry.
+
+        This provides a generic-type-checking alternative to the `@assessment` decorator.
+        """
+        cls._assessment_registry[f.__name__] = f
+        return f
+
+    def eval_cases(self) -> list[EvaluationRow[InputsT, OutputT, MetadataT]]:
+        registry = self.assessment_registry()
+
+        evaluation_rows: list[EvaluationRow[InputsT, OutputT, MetadataT]] = []
+        errors: list[ValueError] = []
+        row_type = self._row_type()
+        for row in self.rows:
+            assessments: list[Assessment[Any, Any, Any]] = []
+            for spec in row.assessments + self.default_assessments:
+                try:
+                    assessment = Assessment[InputsT, OutputT, MetadataT].from_registry(registry, row.name, spec)
+                except ValueError as e:
+                    errors.append(e)
+                    continue
+                assessments.append(assessment)
+            evaluation_rows.append(
+                row_type(
+                    name=row.name,
+                    inputs=row.inputs,
+                    metadata=row.metadata,
+                    expected_output=row.expected_output,
+                    assessments=assessments,
+                )
+            )
+        if errors:
+            raise ExceptionGroup('Failed to load assessments from registry', errors)
+        return evaluation_rows
+
+    # TODO: Task: Always save a schema file when saving the dataset
+    def save(self, path: Path | str = DEFAULT_DATASET_PATH, schema_ref: str | None = None) -> None:
+        path = self._get_relative_path(path)
+        content = yaml.dump(to_jsonable_python(self), sort_keys=False)
+        if schema_ref is not None:
+            content = _ensure_yaml_language_server_line(content, schema_ref)
+        path.write_text(content)
+
+    @classmethod
+    def from_yaml(cls, path: Path | str = DEFAULT_DATASET_PATH) -> Self:
+        path = cls._get_relative_path(path)
+        if not path.exists():
+            raise FileNotFoundError(f'{cls.__name__} dataset file {path} does not exist')
+
+        raw = path.read_text()
+        loaded = yaml.safe_load(raw)
+        try:
+            result = cls.model_validate(loaded)
+            # result.rows = [cls.serialized_row_type().model_validate(row.model_dump()) for row in result.rows]
+        except ValidationError as e:
+            raise ValueError(
+                f'{cls.__name__} dataset file {path} contains data that does not match the schema:\n{e}.'
+            ) from e
+        return result
```

Comment:
```suggestion
        try:
            path = cls._get_relative_path(path)
            raw = path.read_text()
            loaded = yaml.safe_load(raw)
            return cls.model_validate(loaded)
        except Exception as e:
            raise ValueError(f'Failed to load {cls.__name__} from dataset file {path}') from e
```

Reply:
while I would agree this is a lot cleaner for dev, I think the error messages are worse for users. In particular, I don't want to have to scroll up to see the validation errors if they are present.